### PR TITLE
allow setting and deleting of attributes on supported XML::Node types

### PR DESCRIPTION
**Pull request by bmmcginty**
Allow node[name]=value and node.attributes.delete(name)

I'm not sure if Attributes#delete is the correct place for the delete method, but I'm open to suggestions. Maybe Node#delete_attribute?
Also, are we avoiding all methods in libxml2 that require the LIBXML_TREE_ENABLED define? (I've continued that pattern in this PR.)
That should probably be noted somewhere if so.

### CODE REVIEW

**Comment by asterite** [comment]
Thank you for this!

I initially wrote the libxml2 bindings and I mainly focused on parsing XML, not so much in modifying it afterwards. So that the code is not using functions that depend on `LIBXML_TREE_ENABLED` is just a coincidence. I would guess that flag is true by default in most systems, so don't hesitate to use any kind of function from libxml2.

Some comments:
* What do you think if we make `node["attr"] = nil` remove the attribute? It seems reasonable to me, and that way one can pass a value that could be String or Nil and have the attribute set or removed based on that.
* I wouldn't raise an exception when trying to delete an attribute that doesn't exist, just like deleting a key from a Hash, when the key doesn't exist, doesn't raise

File: src/xml/attributes.cr [hunk 1]
```diff
@@ -37,6 +37,23 @@ struct XML::Attributes
...
+    if curr.null?
+      raise KeyError.new("Missing attribute: #{name}")
+    end
+    LibXML.xmlRemoveProp(curr)
+  end
+
+  def []=(name : String, value : String)
+    curr = self[name]?
+    if curr
+      curr.not_nil!.content = value
```
Comment (asterite) [inline src/xml/attributes.cr]: This `not_nil!` is not needed because you already checked that it's not nil with `if curr` above

**Comment by bmmcginty** [comment]
No more raising and Node[attr]=nil now deletes.
I can try rebasing to clean this up if all looks well.

**Comment by asterite** [comment]
Hi @bmmcginty ,

Sorry for what I'm going to say, but we had a discussion about this with @bcardiff and @spalladino and we finally decided that:
* `node[name] = value` could look really weird if `value` might be `nil` and the attribute is deleted, because that looks like an assignment (for example doing that in a Hash will always insert that key/value pair)
* `node.delete(name)` and `attributes.delete(name)` is better than the above
* restrict the `[]=` method to only accept String as value (like the code you proposed, but without the `Nil` overload I proposed)

In Nokogiri you can assign any type as the value, for example `1` or `nil`, and it ends up converting that value to string and assigning it. But it's a bit weird to do `node[name] = nil` and then having `node[name]` be the empty string, so we are not sure about copying that behaviour. If you want you can open an issue to discuss that, but it's low priority.

In short... could you revert your commits to your original proposal? 😊 

Thank you!

**Comment by bmmcginty** [comment]
No problem. That completely makes sense.

**Comment by bmmcginty** [comment]
Fixed. []=(string,nil) removed. node.delete calls attributes.delete. attributes.delete deletes attribute if exists, and returns the nodes content or nil if non-existing attribute.

File: src/xml/node.cr [hunk 2]
```diff
@@ -522,4 +534,4 @@ struct XML::Node
     ptr = @node.value._private
     ptr ? (ptr.as(Array(XML::Error))) : nil
   end
-end
+end
```
Comment (Sija) [inline src/xml/node.cr]: No newline at the end of file.

**Comment by bmmcginty** [comment]
@Sija added blank line. travis builds seem to be timing out, though I don't have a Mac to test on.

**Comment by RX14** [comment]
@bmmcginty the linux builds are failing due to a formatter error

**Comment by Sija** [comment]
@bmmcginty I'd suggest running `crystal format` to have it all fixed in one go.

**Comment by bmmcginty** [comment]
@Sija Format has been run and code rebased. Error appears to be just that odd mac one on Travis. Thanks for your patience.

**Comment by RX14** [comment]
@bmmcginty CI on OSX has been fixed on master, if you rebase your PR onto master it should be fixed.

**Comment by bmmcginty** [comment]
@RX14 rebased, pushed, and passed. thanks.

File: src/xml/node.cr [hunk 3]
```diff
@@ -21,19 +21,29 @@ struct XML::Node
...
   end
 
   # Gets the attribute content for the *attribute* given by name.
-  #
   # Returns `nil` if attribute is not found.
   def []?(attribute : String) : String?
     attributes[attribute]?.try &.content
   end
 
+  # sets *attribute* of this node to *value*. Raises `XML::Error` if this node does not support attributes.
```
Comment (Sija) [inline src/xml/node.cr]: `sets` -> `Sets`

File: src/xml/node.cr [hunk 4]
```diff
@@ -21,19 +21,29 @@ struct XML::Node
   end
 
   # Gets the attribute content for the *attribute* given by name.
-  #
```
Comment (Sija) [inline src/xml/node.cr]: what's the reason for removal?

File: src/xml/node.cr [hunk 5]
```diff
@@ -21,19 +21,29 @@ struct XML::Node
...
   # Raises `KeyError` if attribute is not found.
   def [](attribute : String) : String
     attributes[attribute].content || raise(KeyError.new("Missing attribute: #{attribute}"))
   end
 
   # Gets the attribute content for the *attribute* given by name.
-  #
```
Comment (Sija) [inline src/xml/node.cr]: ditto

File: src/xml/node.cr (hunk 3 above)
Comment (Sija) [inline src/xml/node.cr]: Double newline before 2nd sentence would do good.

File: src/xml/node.cr [hunk 6]
```diff
@@ -21,19 +21,29 @@ struct XML::Node
...
   end
 
+  # sets *attribute* of this node to *value*. Raises `XML::Error` if this node does not support attributes.
+  def []=(name : String, value : String)
+    raise XML::Error.new("Can't set attribute of #{type}", 0) unless element?
+    attributes[name] = value
+  end
+
+  # Deletes attribute given by *name*.
+  # Returns attributes value, or `nil` if attribute not found.
```
Comment (Sija) [inline src/xml/node.cr]: Extra newline in between would do good. See previous comments.

Comment (bcardiff) [inline src/xml/node.cr]: @Sija there is no consistency regarding if the docs for the returned / raised should be in a new paragraph. Applying your proposed changes will left xml/node docs inconsistent.

Let's merge this as it is and unify formatting later. I agree a new line looks better in the html rendering but in an .cr file I would probably forgot to include it.

**Comment by bcardiff** [comment]
Thanks @bmmcginty

**Comment by joenas** [comment]
Is it just me or has there never been a `XML::Attributes#delete`?

```
require "xml"
xml = XML.parse("<div></div>")
xml.delete("test")
=> undefined method 'delete' for XML::Attributes
```

**Comment by RX14** [comment]
@joenas looks like a bug, can you report it?

**Comment by asterite** [comment]
There's unlink

**Comment by joenas** [comment]
@RX14 absolutely, issue here or elsewhere?

**Comment by joenas** [comment]
I’ll see what I can do :) spending most of my free time porting `readability` to crystal atm

**Comment by bmmcginty** [comment]
If you haven't done this yet, let me know and I'll do it this afternoon. (I might be responsible for this myself.)

**Comment by joenas** [comment]
@bmmcginty Sorry I just saw this, I'm i CET so... anyways I haven't so feel free!

**Comment by joenas** [comment]
@bmmcginty Actually I need this to get my `readability` port working and it looks like I managed to figure it out... so I could put up a PR if you like

**Comment by bmmcginty** [comment]
Please do. Thanks.

**Comment by joenas** [comment]
Done: #6910
